### ISSUE TITLE
Make SFML dynamically support several VideoBackends for any OS

### ISSUE THREAD

**Issue by substring** [SFML/SFML#2256]
Hi!

Now that Linux at least has 2 possible video backends (X11 and DRM), SFML lacks the possibility to instantiate one or another since it's for now impossible to dynamically determine the video backend. Several points are to consider here :
- have SFML decide at build time which backends are possible, which means compiling both DRM and X11
- make so that SFML will decide at run time which backend to use
- add an environment variable to force SFML to use a backend

As SFML is right now, you compile 1 backend or another, which is really limiting usability of a seamless video backend SFML. This is also something important to consider for the future, and other OSes. The first example that comes to my mind: one day or another you'll have to add the Wayland backend, there will hardly be anyway to escape it in the coming years.

My simplistic brain has an idea to implement that, but I'm far from a C++ black belt (video backends inherit from a parent abstract class, and have a static method `isAvailable()` so SFML can determine at runtime if a backend is available and decide which way to go)

**Comment by eXpl0it3r** [comment]
Can you add some common use cases for this?

I can potentially imagine some, but as I'm not a Linux user, I'm not quite sure I fully understand the scope in which this is needed. 🤔

**Comment by substring** [comment]
Here is a very simple use case : write a game with SFML that can build on SBCs like a Raspberry Pi where KMS is a *very* popular backend, although X still exists (and Wayland as well). You build a single binary and want that binary to work on both X and KMS. As of today, this is not possible for 2 reasons:
- when building SFML for Linux, you have either X support, either KMS support, not both. 
- the SFML internals aren't ready for supporting several backends and deciding which way to go ,and that's the aim of this issue

**Comment by ChrisThrasher** [comment]
> The first example that comes to my mind: one day or another you'll have to add the Wayland backend, there will hardly be anyway to escape it in the coming years.

Why can't that just be yet another configuration option like the DRM backend is? I don't see why a Wayland backend wouldn't fit into how we're already handing X11 and DRM.

> Wouldn't this partially work against the reason the DRM implementation was added, so one could build SFML without having to link Xlib?

Exploiter is right. I don't want users installing multiple backends just to use one, as will be the use case for most. I also don't like the idea of conditionally compiling in all the backends which are found and ignoring those which are not installed.

I'm sorry but I'm not in favor of this. I think it adds more overhead than it adds value.

**Comment by substring** [comment]
Please read my example. You would force any SFML app to stick to a video a backend when the aim of this issue is to have it adapt to the environment at runtime. It also mean that devs would provide a binary for X and a binary for KMS. That's not really user firendly. That SFML links X libs when the app is started in a DRM context is not a big deal really, and shouldn't be considered as a hurdle to decide about this issue.

Now of course I've started looking at the code and how to do it. Right now the video backend is a compile-time typedef, I do know that selecting the video backend at runtime is a big change. I have a few ideas, and I do know this means adapting other OSes video backends too

> Exploiter is right. I don't want users installing multiple backends just to use one, as will be the use case for most. I also don't like the idea of conditionally compiling in all the backends which are found and ignoring those which are not installed.

Distros compile SFML the way they want, so do power users, that's not something you should worry about. DRM is ALWAYS shipped with linux as it is now a core element.

**Comment by ChrisThrasher** [comment]
> Distros compile SFML the way they want, so do power users, that's not something you should worry about.

I'm worried about the extra work I will have to do to maintain a substantially more complicated build system as the person who spends the most time on build system and tooling code.

**Comment by eXpl0it3r** [comment]
> That SFML links X libs when the app is started in a DRM context is not a big deal really, and shouldn't be considered as a hurdle to decide about this issue.

Except if Xlib is not installed on the target system, no?

Or are you suggesting that Xlib is dynamically loaded (i.e. dlopen)?

**Comment by substring** [comment]
There are many parts of X installed since libGL.so uses a few X libs. No distro (and probably hardly anyone) builds OpenGL without linking libX11 libxcb libXau libXdmcp.

dlopen is a good way to avoid linking X libs. It seems to me you already do it with Vulkan

**Comment by substring** [comment]
This issue is 2 years old now. Are features for 3.0.0 frozen ? Please consider this. If I can help lifting some uncertainties:

- For the easy part -> building:
  - cmake should detect what is available and eventually have some switch to force/disable some backends. This shouldn't be a big deal considering what has been done so far
  - now how distros package is up to them. Cmake is anyway ready. But considering DRM is availabe anywhere, very few distro would manually disable it in SFML
- forget Wayland as of now, there are wrappers for X
- the one and foremost blocker is having SFML being able to detect at runtime what backend is available. X first, then DRM if X is not available. So yes, that's the most complicated thing I agree. Maybe a way is to have some static methods from video backends that return if that backend is available without even initiating it. Anyway, if X is not available, you're good to fallback to DRM.

Once again, to make it clear why I'm pinging that feature: an app shouldn't care about the available video backend and simply ask SFML a window. Whether it's X or DRM is SFML's concern, not the app's. That way, any app can run on X or DRM on whichever Linux version seamlessly.

**Comment by ChrisThrasher** [comment]
> Are features for 3.0.0 frozen ?

Yes, we stopped accepting new features sometime this past summer. There are a number of features we have targeted for v3.1 and will resume work on them once v3 ships.
